torch/csrc/jit/frontend/concrete_module_type.cpp

```diff
@@ -57,15 +44,43 @@ ClassTypePtr ConcreteModuleTypeBuilder::createTypeFromThis() const {
 
 std::shared_ptr<ConcreteModuleType> ConcreteModuleType::fromJitType(
     TypePtr type) {
+  ConcreteModuleTypeBuilder builder;
+  builder.setPoisoned();
+
   // `type` should either be a module interface or a class type
   if (auto interface = type->cast<InterfaceType>()) {
     TORCH_INTERNAL_ASSERT(interface->is_module());
   } else {
-    TORCH_INTERNAL_ASSERT(type->cast<ClassType>());
+    const auto classType = type->expect<ClassType>();
+
+    // Populate the builder metadata from the JIT type. This is to ensure
+    // ConcreteModuleTypes produced from Python and ones produced from a JIT
+    // type directly behave the same to the rest of the system.
+    for (size_t i = 0; i < classType->numAttributes(); i++) {
```

Comment:
I think this makes the code a little less readable for no clear benefit, so I'm going to skip this one unless you feel strongly about it.

Reply:
I prefer my suggestion to avoid making an extra function call every time the loop exit condition is tested: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

But `numAttributes()` is not an expensive call, so I will not lose sleep over this.

Reply:
yes, I would expect a compiler to inline and then avoid recomputing numAttributes(). I think the suggestion makes more sense if we are computing the iterator of a complicated data structure instead of returning an int. Although if the data structure is that complicated, writing a raw for loop over its iterator is probably bad to begin with…